### DISCUSSION TITLE
string slicing and builtin methods clone strings unnecessarily

### DISCUSSION THREAD

**Discussion by fantassin** [vlang/v#11687] [Ideas]
I would like to propose that V slices are implemented as 'views'.

Consider this code;
```
arr := [1,2,3,4,5]
slc := arr[1..3] // This will create a copy instead of simply referring to arr
```
**Edit by `@ntrel`:** The above example is wrong, arrays do not copy data. String slicing does copy data though.

I would suggest that this behavior is changed (or some other syntax is proposed) where slicing simply creates a view into the original chunk of memory without any allocation or copying of elemens.

Motivation:

Consider a simple text/csv file containing one geo point per line expressed as a latitude and a longitude. Here is an example of the structure;
```
-57.9107,99.6348
-0.370774,-58.5762
32.2507,104.6
73.9673,125.027
-15.3535,128.195
72.2073,57.0096
-31.0057,145.352
84.8589,-1.44776
…
```

I used the following V program to create 10 million points;

```
import os
import rand

fn main() {
        mut file := os.create('geo_points.txt') or {
                println('Error')
                return
        }
        for _ in 0 .. 10000000 {
                lat := rand.f64_in_range(-90, 90)
                lon := rand.f64_in_range(-180, 180)
                file.writeln('$lat,$lon')
        }
}
```

Now suppose you have to compute the centroid which is simply the sum of all latitudes and the sum of all longitudes divided by the number of points respectively.

Here is a naive implementation in V;
```
import os

fn main()
{
        s := os.read_file('geo_points.txt') or { println('error') return }

        mut lat_tot := f64(0)
        mut lon_tot := f64(0)
        mut lines := i64(0)

        for line in s.split_into_lines()
        {
                lines++
                fields := line.split(',')
                lat_tot += fields[0].f64()
                lon_tot += fields[1].f64()
        }

        println('Total number of points: $lines')
        println('Centroid { lat: ${lat_tot/lines}, lon: ${lon_tot/lines} }')
}
```
And here is a naive implementation in D;
```
import std.stdio;
Import std.mmfile;
import std.algorithm.iteration;
import std.conv;

void main()
{
        auto mmf = new MmFile("geo_points.txt");
        auto size = mmf.length;
        auto s = cast(string)mmf[];

        double lat_tot = 0;
        double lon_tot = 0;
        ulong line_cnt = 0;

        foreach (ref line; s.splitter('\n'))
        {
                ++line_cnt;

                bool lat = true;
                foreach(ref field; line.splitter(','))
                {
                        if (lat)
                        {
                                lat_tot += to!double(field);
                                lat = false;
                                continue;
                        }

                        lon_tot += to!double(field);
                }
        }

        double centroid_lat = lat_tot / line_cnt;
        double centroid_lon = lon_tot / line_cnt;

        writefln("Total number of points: %d", line_cnt);
        writefln("Centroid { lat: %f, lon: %f }", centroid_lat, centroid_lon);
}
```

I compiled the two versions in release mode and ran over the previously created file.
But first I ran wc to bring all the data in cache, here is the output;

```
(ldc-1.22.0)root@minio ~/slice_test # time wc geo_points.txt
 10000000  10000000 167954804 geo_points.txt

real    0m0.977s
user    0m0.965s
sys     0m0.012s

// THIS THE D VERSION
(ldc-1.22.0)root@minio ~/slice_test # time ./centroid_d
Total number of points: 10000001
Centroid { lat: 0.027982, lon: 0.001638 }

real    0m1.416s
user    0m1.411s
sys     0m0.004s

// THIS IS THE V VERSION
(ldc-1.22.0)root@minio ~/slice_test # time ./centroid_v
Total number of points: 10000000
Centroid { lat: 0.0279825, lon: 0.00163757 }

real    0m7.004s
user    0m6.544s
sys     0m0.460s
```

As you can observe the V’s generated code is almost exactly 5 times slower.
This is totally predictable since V’s slices allocate memory and copy instead of simply present a view.

Let’s run valgrind;

// FIRST D
```
(ldc-1.22.0)root@minio ~/slice_test # valgrind ./centroid_d
==17752== Memcheck, a memory error detector
==17752== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==17752== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==17752== Command: ./centroid_d
==17752==
Total number of points: 10000001
Centroid { lat: 0.027982, lon: 0.001638 }
==17752==
==17752== HEAP SUMMARY:
==17752==     in use at exit: 56 bytes in 2 blocks
==17752==   total heap usage: 549 allocs, 547 frees, 197,220 bytes allocated
==17752==
==17752== LEAK SUMMARY:
==17752==    definitely lost: 0 bytes in 0 blocks
==17752==    indirectly lost: 0 bytes in 0 blocks
==17752==      possibly lost: 32 bytes in 1 blocks
==17752==    still reachable: 24 bytes in 1 blocks
==17752==         suppressed: 0 bytes in 0 blocks
==17752== Rerun with --leak-check=full to see details of leaked memory
==17752==
==17752== For counts of detected and suppressed errors, rerun with: -v
==17752== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

// THEN V
```
(ldc-1.22.0)root@minio ~/slice_test # valgrind ./centroid_v
==17753== Memcheck, a memory error detector
==17753== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==17753== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==17753== Command: ./centroid_v
==17753==
Total number of points: 10000000
Centroid { lat: 0.0279825, lon: 0.00163757 }
==17753==
==17753== HEAP SUMMARY:
==17753==     in use at exit: 1,112,299,946 bytes in 60,000,003 blocks
==17753==   total heap usage: 60,000,048 allocs, 45 frees, 1,548,708,791 bytes allocated
==17753==
==17753== LEAK SUMMARY:
==17753==    definitely lost: 776,390,338 bytes in 30,000,003 blocks
==17753==    indirectly lost: 335,909,600 bytes in 29,999,999 blocks
==17753==      possibly lost: 8 bytes in 1 blocks
==17753==    still reachable: 0 bytes in 0 blocks
==17753==         suppressed: 0 bytes in 0 blocks
==17753== Rerun with --leak-check=full to see details of leaked memory
==17753==
==17753== For counts of detected and suppressed errors, rerun with: -v
==17753== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

We can ignore the leaks since that work is still in progress but the actual allocations are not going away, sadly.

D makes 549 allocations and those have to do with initializing the D runtime which is much heavier than V’s simply because D has a GC.
V makes 60,000,048 allocations which is what kills the performance. 
NOTE: To be clear a lot of the allocations come from parsing the strings into numbers and not just creating strings.

I could probably write a dissertation on why memory allocation on the hot path is a terrible idea but the main pain points are;
 - Overhead, if you allocating memory then you are not working on your problem at hand
 - Side effect, allocations (unless using highly specialized heaps) require memory synchronization which in turn impacts code in 
   your process that has nothing to do with the code making the allocation
 - Memory fragmentation, this is one of the worst effects since it impacts reliability and performance
 - Point of failure, this is probably the most impactful when reliable software is to be written since it’s practically impossible to 
   recover from a failure to allocate.

Hopefully this will convince you and open the discussion for slices as views in V.

**Note: 06/26/2020**

For what it's worth, I re-wrote the test in V again in C style (emulating slices):

Compiled with GCC:
```
root@minio ~/slice_test # time ./a.out
Total number of points: 9999999
Centroid { lat: 0.0279842, lon: 0.00162253 }

real    0m1.340s
user    0m1.327s
sys     0m0.012s
```
This is almost 10% faster than the D version.

Compiled with Clang:
```
root@minio ~/slice_test # time ./a.out
Total number of points: 9999999
Centroid { lat: 0.0279842, lon: 0.00162253 }

real    0m0.653s
user    0m0.641s
sys     0m0.012s
```

Now this is **more than two times** faster! I haven't had time to try to understand why.

Valgrind shows 19 allocs.
```
root@minio ~/slice_test # valgrind ./a.out
==24199== Memcheck, a memory error detector
==24199== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==24199== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==24199== Command: ./a.out
==24199==
Total number of points: 9999999
Centroid { lat: 0.0279842, lon: 0.00162253 }
==24199==
==24199== HEAP SUMMARY:
==24199==     in use at exit: 12,221 bytes in 16 blocks
==24199==   total heap usage: 19 allocs, 3 frees, 13,501 bytes allocated
==24199==
==24199== LEAK SUMMARY:
==24199==    definitely lost: 77 bytes in 2 blocks
==24199==    indirectly lost: 0 bytes in 0 blocks
==24199==      possibly lost: 0 bytes in 0 blocks
==24199==    still reachable: 12,144 bytes in 14 blocks
==24199==         suppressed: 0 bytes in 0 blocks
==24199== Rerun with --leak-check=full to see details of leaked memory
==24199==
==24199== For counts of detected and suppressed errors, rerun with: -v
==24199== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by krolaw** [comment]
Sometimes it is desirable to edit a large array by slicing the part to edited and passing that to a fn.

**Comment by ntrel** [comment]
> This is almost 10% faster than the D version.
> ...
> Now this is more than two times faster! I haven't had time to try to understand why.

You are using a memory-mapped file in D when you should use `std.file.read`.
Also, although D doesn't need to allocate a `fields` array (you can use `front` and `popFront` instead of `foreach` ;-)), for equivalence I would allocate using `splitter(',').array`:
https://gist.github.com/ntrel/1958c700ad0af9c05e3a12cf61cfea6b

*Edit:* `read` avoids the Unicode validation of `readText`.

**Comment by fantassin** [comment]
@ntrel Thank you for the feedback.

However my post was about slicing of arrays and how they impact performance. I tried to come up with an implementation that is still idiomatic D and V without sacrificing performance. Your version is almost two times slower than my D implementation, so it is of no interest;
```
(ldc-1.22.0)root@minio ~/slice_test # time ./centr
Total number of points: 10000000
Centroid { lat: 0.027982, lon: 0.001638 }

real    0m2.605s
user    0m2.586s
sys     0m0.048s
```

>You are using a memory-mapped file in D when you should use std.file.read.

I actually don't want to do that since I am interested in performance, not equivalency with V. V as of yet does not have a cross platform ```mmap()``` ASFAIK, so I used ```read()```. But in my custom implementation that is 2 times faster than the D version I do use ```mmap()```.

**Comment by medvednikov** [comment]
There will be an option to do views for strings and slices.

**Comment by fantassin** [comment]
@medvednikov That is awesome!
Is the syntax decided or still being discussed?
It would be great if the distinction between slices that simply make a copy vs non-owning slices as views is clear and obvious in V code.

**Comment by ntrel** [comment]
> I am interested in performance, not equivalency with V

@fantassin In that case I suggest you don't read the whole file into memory, just the current line:
```d
        foreach (line; File("geo_points.txt").byLine)
        {
                ++line_cnt;
                auto fields = line.splitter(',');
                lat_tot += fields.front.to!double;
                fields.popFront;
                lon_tot += fields.front.to!double;
        }
```

**Comment by ntrel** [comment]
> // This will create a copy instead of simply referring to arr
> arr := [1,2,3,4,5]
> slc := arr[1..3] // This will create a copy instead of simply referring to arr

That is not true, look what happens when we mutate the slice:
```v
mut arr := [1,2,3,4,5]
mut slc := arr[1..3]
slc[0] = -2
assert arr[1] == -2
```
The problem with strings is that `str.substr` creates a copy, and `substr` is called a lot in `builtin/string.v`. I raised this as an issue here: https://github.com/vlang/v/issues/1269. At the end I think Alex says he intends to only clone the data when a string escapes its owner's scope.

**Comment by fantassin** [comment]
Closing this as invalid.

**Comment by ntrel** [comment]
I don't think this is invalid, because it describes a problem that unnecessary copies are made in string methods. Reopening.

**Comment by fantassin** [comment]
@ntrel I think the issue becomes the inconsistency between slicing arrays and strings. Arrays behave correctly (in the sense that slices are non-owning views) whereas slicing strings (using the same operator [a .. b]) invokes substr() which makes a new string.

I think that `s[a .. b]` where `s` is a string should return a view just like arrays and an explicit `substr()` method should be used when a copy is needed. That would basically resolve this issue. wdyt?

**Comment by ntrel** [comment]
@fantassin
> slicing strings (using the same operator [a .. b]) invokes substr()

OK, thanks.
```v
s := 'ok'
s2 := s[0..2]
assert s.str == s2.str //fails
```
> an explicit substr() method should be used when a copy is needed.

Edit: I think `clone()` should be used when a copy is needed.

**Comment by ntrel** [comment]
Since last summer builtin/string.v has changed to require null terminated string data:
```v
	str byteptr // points to a C style 0 terminated string of bytes.
	len int // the length of the .str field, excluding the ending 0 byte. It is always equal to strlen(.str).
```
This means you can't slice strings without allocating. This is very bad for performance. We should not be copying C's mistake of null-terminated strings :-(

Instead we should null terminate string literals only. @medvednikov
